### ISSUE TITLE
Why is np.dot imprecise? (n-dim arrays)

### ISSUE THREAD

**Issue by OverLordGoldDragon** [numpy/numpy#14917]
Ported from the [original SO](https://stackoverflow.com/questions/58740925/why-is-np-dot-imprecise-n-dim-arrays) - full info there. Ideally, I seek a _code-level_ explanation of the behavior.

<hr>

Suppose we take `np.dot` of two `'float32'` 2D arrays:  
 
```python
res = np.dot(a, b)   # see CASE 1
print(list(res[0]))  # list shows more digits
``` 
```python
[-0.90448684, -1.1708503, 0.907136, 3.5594249, 1.1374011, -1.3826287]
```
Numbers. Except, they can change:

<hr>

**CASE 1**: slice `a`
```python
np.random.seed(1)
a = np.random.randn(9, 6).astype('float32')
b = np.random.randn(6, 6).astype('float32')

for i in range(1, len(a)):
    print(list(np.dot(a[:i], b)[0])) # full shape: (i, 6)
```
```
[-0.9044868,  -1.1708502, 0.90713596, 3.5594249, 1.1374012, -1.3826287]
[-0.90448684, -1.1708503, 0.9071359,  3.5594249, 1.1374011, -1.3826288]
[-0.90448684, -1.1708503, 0.9071359,  3.5594249, 1.1374011, -1.3826288]
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]
```
Results differ, even though the printed slice derives from the exact same numbers multiplied.
<hr>
**CASE 2**: flatten `a`, take a 1D version of `b`, _then_ slice `a`:
```python
np.random.seed(1)
a = np.random.randn(9, 6).astype('float32')
b = np.random.randn(1, 6).astype('float32')

for i in range(1, len(a)):
    a_flat = np.expand_dims(a[:i].flatten(), -1) # keep 2D
    print(list(np.dot(a_flat, b)[0])) # full shape: (i*6, 6)
```
```python
[-0.3393164, 0.9528787, 1.3627989, 1.5124314, 0.46389243, 1.437775]
[-0.3393164, 0.9528787, 1.3627989, 1.5124314, 0.46389243, 1.437775]
[-0.3393164, 0.9528787, 1.3627989, 1.5124314, 0.46389243, 1.437775]
[-0.3393164, 0.9528787, 1.3627989, 1.5124314, 0.46389243, 1.437775]
[-0.3393164, 0.9528787, 1.3627989, 1.5124314, 0.46389243, 1.437775]
[-0.3393164, 0.9528787, 1.3627989, 1.5124314, 0.46389243, 1.437775]
[-0.3393164, 0.9528787, 1.3627989, 1.5124314, 0.46389243, 1.437775]
[-0.3393164, 0.9528787, 1.3627989, 1.5124314, 0.46389243, 1.437775]
```
<hr>

**CASE 3**: stronger control; set all non-involved entires to _zero_: add `a[1:] = 0` to CASE 1 code. Result: discrepancies persist.

<hr>

**CASE 4**: check indices other than `[0]`; like for `[0]`, results begin to stabilize a fixed # of array enlargements from their point of creation. [Output](https://pastebin.com/wwCYiKxp)

```python
np.random.seed(1)
a = np.random.randn(9, 6).astype('float32')
b = np.random.randn(6, 6).astype('float32')

for j in range(len(a) - 2):
    for i in range(1, len(a)):
        res = np.dot(a[:i], b)
        try:    print(list(res[j]))
        except: pass
    print()
```

<hr>

Hence, for the 2D * 2D case, results differ - but are consistent for 1D * 1D. From some of my readings, this appears to stem from 1D-1D using simple addition, whereas 2D-2D uses 'fancier', performance-boosting addition that may be less precise (e.g. pairwise addition does the opposite). Nonetheless, I'm unable to understand why discrepancies vanish in case 1 once `a` is sliced past a set 'threshold'; the larger `a` and `b`, the later this threshold seems to lie, but it always exists.

All said: why is `np.dot` imprecise (and inconsistent) for ND-ND arrays?

<hr>

**Additional info**:

 - <b>Environment</b>: Win-10 OS, Python 3.7.4, Spyder 3.3.6 IDE, Anaconda 3.0 2019/10
 - <b>CPU</b>: [i7-7700HQ 2.8 GHz](https://ark.intel.com/content/www/us/en/ark/products/97185/intel-core-i7-7700hq-processor-6m-cache-up-to-3-80-ghz.html)
 - <b>Numpy</b> v1.16.5

**Possible culprit library**: Numpy [MKL](https://docs.anaconda.com/mkl-optimizations/) - also BLASS libraries; thanks to [Bi Rico](https://stackoverflow.com/users/1004096/bi-rico) for noting

<hr>

**Stress-test code**: as noted, discrepancies exacerbate in frequency w/ larger arrays; if above isn't reproducible, below should be (if not, try larger dims). [My output](https://pastebin.com/kGFY3vwZ)

```python
np.random.seed(1)
a = (0.01*np.random.randn(9, 9999)).astype('float32') # first multiply then type-cast
b = (0.01*np.random.randn(9999, 6)).astype('float32') # *0.01 to bound mults to < 1

for i in range(1, len(a)):
    print(list(np.dot(a[:i], b)[0]))
``` 

<hr>

**Problem severity**: shown discrepancies are 'small', but no longer so when operating on a neural network with billions of numbers multiplied over a few seconds, and trillions over the entire runtime; reported model accuracy differs by entire 10's of percents, per [this thread](https://github.com/keras-team/keras/issues/12400). 

Below is a gif of arrays resulting from feeding to a model what's basically `a[0]`, w/ `len(a)==1` vs. `len(a)==32`:

<img src="https://i.stack.imgur.com/nOqIG.gif" width="350">

**Comment by shoyer** [comment]
NumPy does not (and cannot) guarantee that two mathematically equivalent operations will give exactly the same result on floating point numbers. Floating point errors should be expected, especially in single precision (float32). They are a fundamental limitation of floating point arithmetic.

For a simpler example without even using NumPy, note:
```
>>> (0.1 + 0.2) + 0.3 == 0.1 + (0.2 + 0.3)
False
```

Operations like `np.dot` in particular are highly optimized, and use different code paths for different array shapes. The precise rounding behavior is not guaranteed to be consistent, because that would make the implementations slower.

The bottom line is that if you need accuracy in the last decimal point, you should use higher precision numbers, e.g., float64 in this case.

**Comment by OverLordGoldDragon** [comment]
@shoyer Yes, that much's been clarified in the SO - but my question is, which _specific algorithm / code_ is used by `np.dot`? If that's not to be answered here, where can I learn about it? The code's highly encapsulated and branches into different libraries (BLAS, MKL, etc), so it's not of much help.

**Comment by mattip** [comment]
What do you mean "not of much help"? There is no high-level flow diagram of the code, and if there was it would be (in my opinion) hopelessly more complicated than the code itself. If you want to know what specific code/algorithm is used for each of 1d.dot(2d), 1d.dot(1d), 2d.dot(2d), 2d.dot(1d) for any particular pair of dtypes, I can think of no better path than to dive into the code.

I would start at `PyArray_MatrixProduct2`, which in the general case, given a CBLAS backend, will call `cblas_matrixproduct` for BLAS-supported dtypes and will use an iterator with the output dtype's `dotfunc` otherwise.

**Comment by OverLordGoldDragon** [comment]
@mattip "I would start at" -- exactly what I'm looking for, _code specifics_. Though, the moment we hit C, code complexity skyrockets - and good luck sorting through that; for one, `PyArray_MatrixProduct2` is a wall of imports, among which `cblas_matrixproduct` or `CBLAS` can't be found via Ctrl + F (it's probably elsewhere entirely, which is exactly the point).

That order of operations mediates numeric precision I understand, but I mainly seek to understand one specific behavior: "consistent inconsistencies" in increasing dim 0 of `a`:

```python
[-0.9044868,  -1.1708502, 0.90713596, 3.5594249, 1.1374012, -1.3826287]  # <- not same as prev
[-0.90448684, -1.1708503, 0.9071359,  3.5594249, 1.1374011, -1.3826288]  # <- not same as prev
[-0.90448684, -1.1708503, 0.9071359,  3.5594249, 1.1374011, -1.3826288]  # <- not same as prev
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]  # <- not same as prev
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]  # <- same as prev
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]  # <- same as prev
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]  # <- same as prev
[-0.90448684, -1.1708503, 0.907136,   3.5594249, 1.1374011, -1.3826287]  # <- same as prev
```

No matter how much further we increase `a`'s dim 0, subsequent rows are all _consistent_ - but the first few are all over the place. Further, increasing the _inner dimension_ exacerbates discrepancies, and the number of inconsistent rows (this can be understood as more multiplications per row, hence greater chance at error, but unsure). 

If you could point to me to a single specific algorithm that can explain this behavior, I'd consider this resolved.

**Comment by seberg** [comment]
@OverLordGoldDragon the numpy side may be a bit complicated, but it is not even the issue, nor the pinnacle of complexity in any case. This happens inside heavily optimized linear algebra code that is part of OpenBLAS or other BLAS implementations. As you found correctly, it uses `cblas_matrixproduct` which does not do much more than call into BLAS.

**Comment by shoyer** [comment]
You're doing dense matrix/matrix multiplication in single precision, so at the end I'm pretty sure it will boil down to calling `sgemm`. The details of how that is implemented will depend on your BLAS library, and unfortunately you're using Intel's MKL, which is closed source.

**Comment by OverLordGoldDragon** [comment]
Closed source, how unfun; guess it's the dead end. Interestingly, Colab, which uses OpenBLAS, is remarkably consistent - only differing for the first two rows, for within 24 printed digits (didn't try more), and for very large inner dims (`(12, 99999) * (99999, 6)`)

```python
 0.059018023312091827392578 0.016609217971563339233398 -0.007542920764535665512085 -0.013307953253388404846191 0.031583771109580993652344 -0.019641952589154243469238
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
 0.059017922729253768920898 0.016609253361821174621582 -0.007542883045971393585205 -0.013307942077517509460449 0.031583771109580993652344 -0.019641976803541183471680
```
On my local machine, notable differences persist for up to 6 rows - perhaps it's a display of Colab's Intel(R) Xeon(R) @ 2.00GHz superiority to my CPU to be able to afford such precision w/ performance.

I'd still be interested to understand how that first row differs, and I presume OpenBLAS is ... "open" - but if that's also troublesome, suppose I'll call it a day. Either way, thanks for your assistance.

**Comment by nschloe** [comment]
FWIW, accurate (but slow) dot products are in [accupy](https://github.com/nschloe/accupy).
